File: apps/rh/src/lib/planilha/funcionarios.ts
import 'server-only'
import ExcelJS from 'exceljs'
import { apenasDigitos, cpfValido } from '@/lib/dominio/cpf'
import { STATUS } from '@/lib/dominio/constantes'

/**
 * Leitura da planilha de funcionários.
 *
 * Os cabeçalhos são reconhecidos por APELIDO, e não por posição: a planilha de RH de uma
 * construtora nunca vem com as colunas na mesma ordem duas vezes, e exigir um modelo fixo
 * faria o importador ser usado uma vez e abandonado. "Nome", "NOME COMPLETO" e
 * "Funcionário" são a mesma coluna.
 */

const APELIDOS: Record<string, string[]> = {
  nome: ['nome', 'nome completo', 'funcionario', 'colaborador'],
  cpf: ['cpf', 'c.p.f', 'cpf/mf'],
  matricula: ['matricula', 'matrícula', 'registro', 'chapa'],
  rg: ['rg', 'identidade'],
  dataNascimento: ['data de nascimento', 'nascimento', 'dt nascimento', 'data nascimento'],
  admitidoEm: ['admissao', 'admissão', 'data de admissao', 'data de admissão', 'dt admissao', 'admitido em'],
  cargo: ['cargo', 'funcao', 'função'],
  obra: ['obra', 'centro de custo', 'lotacao', 'lotação'],
  telefone: ['telefone', 'celular', 'fone', 'contato'],
  email: ['email', 'e-mail'],
  status: ['status', 'situacao', 'situação'],
  sexo: ['sexo', 'genero', 'gênero'],
  salario: ['salario', 'salário'],
  tipoContrato: ['tipo de contrato', 'contrato', 'regime'],
  cidade: ['cidade', 'municipio', 'município'],
  uf: ['uf', 'estado'],
  tamanhoCamisa: ['camisa', 'tamanho camisa', 'tam camisa'],
  tamanhoCalca: ['calca', 'calça', 'tamanho calca', 'tam calca'],
  tamanhoCalcado: ['calcado', 'calçado', 'bota', 'tamanho calcado', 'tam calcado'],
}

function normalizar(texto: string): string {
  return texto
    .toString()
    .normalize('NFD')
    .replace(/[̀-ͯ]/g, '')
    .toLowerCase()
    .replace(/[^a-z0-9\s]/g, ' ')
    .replace(/\s+/g, ' ')
    .trim()
}

export type LinhaFuncionario = {
  linha: number
  nome: string
  cpf: string
  matricula: string | null
  admitidoEm: Date | null
  cargo: string | null
  obra: string | null
  status: string
  rg: string | null
  dataNascimento: Date | null
  telefone: string | null
  email: string | null
  sexo: string | null
  salario: number | null
  tipoContrato: string | null
  cidade: string | null
  uf: string | null
  tamanhoCamisa: string | null
  tamanhoCalca: string | null
  tamanhoCalcado: string | null
}

export type LinhaIgnorada = { linha: number; motivo: string; conteudo: string }

/**
 * Converte o que veio da célula em data.
 *
 * O Excel devolve `Date` quando a célula está formatada como data e texto quando não está —
 * e a planilha de RH tem as duas coisas na mesma coluna, porque alguém digitou algumas à
 * mão. Aceitar só um dos formatos perderia metade das linhas sem dizer por quê.
 */
function paraData(valor: unknown): Date | null {
  if (!valor) return null

  if (valor instanceof Date) {
    return new Date(Date.UTC(valor.getUTCFullYear(), valor.getUTCMonth(), valor.getUTCDate()))
  }

  const texto = String(valor).trim()
  const br = /^(\d{1,2})[/\-.](\d{1,2})[/\-.](\d{2,4})$/.exec(texto)
  if (br) {
    const ano = br[3].length === 2 ? 2000 + Number(br[3]) : Number(br[3])
    const d = new Date(Date.UTC(ano, Number(br[2]) - 1, Number(br[1])))
    return Number.isNaN(d.getTime()) ? null : d
  }

  const iso = /^(\d{4})-(\d{2})-(\d{2})/.exec(texto)
  if (iso) return new Date(Date.UTC(+iso[1], +iso[2] - 1, +iso[3]))

  return null
}

function texto(valor: unknown): string | null {
  if (valor === null || valor === undefined) return null
  // Célula com fórmula vem como objeto; o que interessa é o resultado.
  const bruto = typeof valor === 'object' && 'result' in (valor as object)
    ? String((valor as { result: unknown }).result ?? '')
    : String(valor)
  const limpo = bruto.trim()
  return limpo ? limpo : null
}

function numero(valor: unknown): number | null {
  const t = texto(valor)
  if (!t) return null
  const n = Number(t.replace(/[R$\s.]/g, '').replace(',', '.'))
  return Number.isFinite(n) ? n : null
}

const STATUS_ACEITOS: Record<string, string> = {
  ativo: STATUS.ATIVO,
  afastado: STATUS.AFASTADO,
  ferias: STATUS.FERIAS,
  desligado: STATUS.DESLIGADO,
  demitido: STATUS.DESLIGADO,
  inativo: STATUS.DESLIGADO,
}

export type LeituraPlanilha = {
  linhas: LinhaFuncionario[]
  ignoradas: LinhaIgnorada[]
  colunasReconhecidas: string[]
  colunasIgnoradas: string[]
}

export async function lerPlanilhaFuncionarios(buffer: ArrayBuffer): Promise<LeituraPlanilha> {
  const wb = new ExcelJS.Workbook()
  await wb.xlsx.load(buffer)

  const aba = wb.worksheets[0]
  if (!aba) throw new Error('A planilha está vazia.')

  // O cabeçalho nem sempre é a primeira linha: é comum haver título e logotipo em cima.
  // Procura nas primeiras dez a linha que tem "nome" e "cpf".
  let linhaCabecalho = 0
  const posicaoDaColuna = new Map<string, number>()

  for (let i = 1; i <= Math.min(10, aba.rowCount); i++) {
    const candidatas = new Map<string, number>()
    aba.getRow(i).eachCell((cell, col) => {
      const t = normalizar(texto(cell.value) ?? '')
      if (!t) return
      for (const [campo, apelidos] of Object.entries(APELIDOS)) {
        if (apelidos.includes(t) && !candidatas.has(campo)) candidatas.set(campo, col)
      }
    })
    if (candidatas.has('nome') && candidatas.has('cpf')) {
      linhaCabecalho = i
      for (const [k, v] of candidatas) posicaoDaColuna.set(k, v)
      break
    }
  }

  if (linhaCabecalho === 0) {
    throw new Error(
      'Não encontrei as colunas "Nome" e "CPF" nas primeiras linhas. Confira se a planilha ' +
        'tem uma linha de cabeçalho com esses nomes.',
    )
  }

  const naoReconhecidas: string[] = []
  aba.getRow(linhaCabecalho).eachCell((cell, col) => {
    const t = texto(cell.value)
    if (!t) return
    if (![...posicaoDaColuna.values()].includes(col)) naoReconhecidas.push(t)
  })

  const ler = (row: ExcelJS.Row, campo: string) => {
    const col = posicaoDaColuna.get(campo)
    return col ? row.getCell(col).value : null
  }

  const linhas: LinhaFuncionario[] = []
  const ignoradas: LinhaIgnorada[] = []
  const cpfsVistos = new Set<string>()

  for (let i = linhaCabecalho + 1; i <= aba.rowCount; i++) {
    const row = aba.getRow(i)
    const nome = texto(ler(row, 'nome'))
    const cpfBruto = texto(ler(row, 'cpf'))

    if (!nome && !cpfBruto) continue // linha em branco: separador, não erro

    const resumo = `${nome ?? '—'} / ${cpfBruto ?? '—'}`

    if (!nome || nome.length < 3) {
      ignoradas.push({ linha: i, motivo: 'sem nome', conteudo: resumo })
      continue
    }

    const cpf = apenasDigitos(cpfBruto ?? '')
    if (!cpf) {
      ignoradas.push({ linha: i, motivo: 'sem CPF', conteudo: resumo })
      continue
    }
    if (!cpfValido(cpf)) {
      ignoradas.push({ linha: i, motivo: 'CPF inválido', conteudo: resumo })
      continue
    }
    if (cpfsVistos.has(cpf)) {
      ignoradas.push({ linha: i, motivo: 'CPF repetido na própria planilha', conteudo: resumo })
      continue
    }
    cpfsVistos.add(cpf)

    const statusBruto = normalizar(texto(ler(row, 'status')) ?? '')
    const admitidoEm = paraData(ler(row, 'admitidoEm'))

    linhas.push({
      linha: i,
      nome,
      cpf,
      matricula: texto(ler(row, 'matricula')),
      admitidoEm,
      cargo: texto(ler(row, 'cargo')),
      obra: texto(ler(row, 'obra')),
      status: STATUS_ACEITOS[statusBruto] ?? STATUS.ATIVO,
      rg: texto(ler(row, 'rg')),
      dataNascimento: paraData(ler(row, 'dataNascimento')),
      telefone: texto(ler(row, 'telefone')),
      email: texto(ler(row, 'email')),
      sexo: texto(ler(row, 'sexo')),
      salario: numero(ler(row, 'salario')),
      tipoContrato: texto(ler(row, 'tipoContrato')),
      cidade: texto(ler(row, 'cidade')),
      uf: texto(ler(row, 'uf')),
      tamanhoCamisa: texto(ler(row, 'tamanhoCamisa')),
      tamanhoCalca: texto(ler(row, 'tamanhoCalca')),
      tamanhoCalcado: texto(ler(row, 'tamanhoCalcado')),
    })
  }

  return {
    linhas,
    ignoradas,
    colunasReconhecidas: [...posicaoDaColuna.keys()],
    colunasIgnoradas: naoReconhecidas,
  }
}

File: migracao-go/internal/infrastructure/planilha/funcionarios.go
// Leitura da planilha de funcionários do RH — ver migracao-go/rh/COMPORTAMENTO.md §5.
// Espelha apps/rh/src/lib/planilha/funcionarios.ts: cabeçalho reconhecido por APELIDO (não
// por posição fixa), diferente do parser do Painel de Locação neste mesmo pacote.
package planilha

import (
	"io"
	"strconv"
	"strings"
	"time"

	"github.com/xuri/excelize/v2"

	dominio "siqueiracampos/servidor/internal/domain/rh"
)

var apelidosFuncionario = map[string][]string{
	"nome": {"nome", "nome completo", "funcionario", "colaborador"},
	// "c p f" e "cpf mf" são "C.P.F"/"CPF/MF" já normalizados — NormalizarCabecalho troca
	// pontuação por espaço, então o apelido precisa estar na forma pós-normalização.
	"cpf":            {"cpf", "c p f", "cpf mf"},
	"matricula":      {"matricula", "registro", "chapa"},
	"rg":             {"rg", "identidade"},
	"dataNascimento": {"data de nascimento", "nascimento", "dt nascimento", "data nascimento"},
	"admitidoEm":     {"admissao", "data de admissao", "dt admissao", "admitido em"},
	"cargo":          {"cargo", "funcao"},
	"obra":           {"obra", "centro de custo", "lotacao"},
	"telefone":       {"telefone", "celular", "fone", "contato"},
	"email":          {"email", "e mail"},
	"status":         {"status", "situacao"},
	"sexo":           {"sexo", "genero"},
	"salario":        {"salario"},
	"tipoContrato":   {"tipo de contrato", "contrato", "regime"},
	"cidade":         {"cidade", "municipio"},
	"uf":             {"uf", "estado"},
	"tamanhoCamisa":  {"camisa", "tamanho camisa", "tam camisa"},
	"tamanhoCalca":   {"calca", "tamanho calca", "tam calca"},
	"tamanhoCalcado": {"calcado", "bota", "tamanho calcado", "tam calcado"},
}

var statusAceitosFuncionario = map[string]string{
	"ativo": dominio.StatusAtivo, "afastado": dominio.StatusAfastado, "ferias": dominio.StatusFerias,
	"desligado": dominio.StatusDesligado, "demitido": dominio.StatusDesligado, "inativo": dominio.StatusDesligado,
}

type LinhaFuncionarioImportado struct {
	Linha                                       int
	Nome, CPF                                   string
	Matricula                                   *string
	AdmitidoEm                                  *time.Time
	Cargo, Obra                                 *string
	Status                                      string
	RG                                          *string
	DataNascimento                              *time.Time
	Telefone, Email, Sexo                       *string
	Salario                                     *float64
	TipoContrato, Cidade, UF                    *string
	TamanhoCamisa, TamanhoCalca, TamanhoCalcado *string
}

type LinhaFuncionarioIgnorada struct {
	Linha    int
	Motivo   string
	Conteudo string
}

type LeituraFuncionarios struct {
	Linhas              []LinhaFuncionarioImportado
	Ignoradas           []LinhaFuncionarioIgnorada
	ColunasReconhecidas []string
	ColunasIgnoradas    []string
}

func vazio(s string) *string {
	s = strings.TrimSpace(s)
	if s == "" {
		return nil
	}
	return &s
}

// LerPlanilhaFuncionarios lê diretamente do upload (io.Reader) — nunca grava em disco, é a
// própria decisão de segurança documentada em COMPORTAMENTO.md §5.
func LerPlanilhaFuncionarios(r io.Reader) (*LeituraFuncionarios, error) {
	f, err := excelize.OpenReader(r)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	folhas := f.GetSheetList()
	if len(folhas) == 0 {
		return nil, errPlanilhaVazia
	}
	sheet := folhas[0]

	linhasBrutas, err := f.GetRows(sheet)
	if err != nil {
		return nil, err
	}
	totalLinhas := len(linhasBrutas)

	// O cabeçalho nem sempre é a primeira linha: procura nas primeiras dez a que tem "nome"
	// e "cpf" entre os apelidos reconhecidos.
	linhaCabecalho := 0
	posicaoDaColuna := map[string]int{}

	limite := totalLinhas
	if limite > 10 {
		limite = 10
	}
	for i := 1; i <= limite; i++ {
		candidatas := map[string]int{}
		for col, valor := range linhasBrutas[i-1] {
			t := dominio.NormalizarCabecalho(valor)
			if t == "" {
				continue
			}
			for campo, apelidos := range apelidosFuncionario {
				if _, ja := candidatas[campo]; ja {
					continue
				}
				for _, a := range apelidos {
					if a == t {
						candidatas[campo] = col + 1 // 1-based, mesma convenção do excelize
						break
					}
				}
			}
		}
		if _, temNome := candidatas["nome"]; temNome {
			if _, temCPF := candidatas["cpf"]; temCPF {
				linhaCabecalho = i
				posicaoDaColuna = candidatas
				break
			}
		}
	}

	if linhaCabecalho == 0 {
		return nil, errCabecalhoNaoEncontrado
	}

	colunasReconhecidas := make([]string, 0, len(posicaoDaColuna))
	for campo := range posicaoDaColuna {
		colunasReconhecidas = append(colunasReconhecidas, campo)
	}

	posicoesUsadas := map[int]bool{}
	for _, col := range posicaoDaColuna {
		posicoesUsadas[col] = true
	}
	var colunasIgnoradas []string
	for col, valor := range linhasBrutas[linhaCabecalho-1] {
		v := strings.TrimSpace(valor)
		if v != "" && !posicoesUsadas[col+1] {
			colunasIgnoradas = append(colunasIgnoradas, v)
		}
	}

	ler := func(row []string, campo string) string {
		col, ok := posicaoDaColuna[campo]
		if !ok || col-1 >= len(row) {
			return ""
		}
		return strings.TrimSpace(row[col-1])
	}

	lerData := func(row []string, campo string, numeroLinha int) *time.Time {
		col, ok := posicaoDaColuna[campo]
		if !ok {
			return nil
		}
		return celulaData(f, sheet, col, numeroLinha)
	}

	var linhas []LinhaFuncionarioImportado
	var ignoradas []LinhaFuncionarioIgnorada
	cpfsVistos := map[string]bool{}

	for i := linhaCabecalho + 1; i <= totalLinhas; i++ {
		row := linhasBrutas[i-1]
		nome := ler(row, "nome")
		cpfBruto := ler(row, "cpf")

		if nome == "" && cpfBruto == "" {
			continue // linha em branco: separador, não erro
		}

		resumo := semTraco(nome) + " / " + semTraco(cpfBruto)

		if len(nome) < 3 {
			ignoradas = append(ignoradas, LinhaFuncionarioIgnorada{Linha: i, Motivo: "sem nome", Conteudo: resumo})
			continue
		}

		cpf := dominio.ApenasDigitos(cpfBruto)
		if cpf == "" {
			ignoradas = append(ignoradas, LinhaFuncionarioIgnorada{Linha: i, Motivo: "sem CPF", Conteudo: resumo})
			continue
		}
		if !dominio.CPFValido(cpf) {
			ignoradas = append(ignoradas, LinhaFuncionarioIgnorada{Linha: i, Motivo: "CPF inválido", Conteudo: resumo})
			continue
		}
		if cpfsVistos[cpf] {
			ignoradas = append(ignoradas, LinhaFuncionarioIgnorada{Linha: i, Motivo: "CPF repetido na própria planilha", Conteudo: resumo})
			continue
		}
		cpfsVistos[cpf] = true

		statusBruto := dominio.NormalizarChave(ler(row, "status"))
		status, ok := statusAceitosFuncionario[statusBruto]
		if !ok {
			status = dominio.StatusAtivo
		}

		linhas = append(linhas, LinhaFuncionarioImportado{
			Linha: i, Nome: nome, CPF: cpf,
			Matricula:      vazio(ler(row, "matricula")),
			AdmitidoEm:     lerData(row, "admitidoEm", i),
			Cargo:          vazio(ler(row, "cargo")),
			Obra:           vazio(ler(row, "obra")),
			Status:         status,
			RG:             vazio(ler(row, "rg")),
			DataNascimento: lerData(row, "dataNascimento", i),
			Telefone:       vazio(ler(row, "telefone")),
			Email:          vazio(ler(row, "email")),
			Sexo:           vazio(ler(row, "sexo")),
			Salario:        celulaNumero(ler(row, "salario")),
			TipoContrato:   vazio(ler(row, "tipoContrato")),
			Cidade:         vazio(ler(row, "cidade")),
			UF:             vazio(ler(row, "uf")),
			TamanhoCamisa:  vazio(ler(row, "tamanhoCamisa")),
			TamanhoCalca:   vazio(ler(row, "tamanhoCalca")),
			TamanhoCalcado: vazio(ler(row, "tamanhoCalcado")),
		})
	}

	return &LeituraFuncionarios{
		Linhas: linhas, Ignoradas: ignoradas,
		ColunasReconhecidas: colunasReconhecidas, ColunasIgnoradas: colunasIgnoradas,
	}, nil
}

func semTraco(s string) string {
	if s == "" {
		return "—"
	}
	return s
}

var (
	errPlanilhaVazia          = planilhaErro("A planilha está vazia.")
	errCabecalhoNaoEncontrado = planilhaErro("Não encontrei as colunas \"Nome\" e \"CPF\" nas primeiras linhas. Confira se a planilha tem uma linha de cabeçalho com esses nomes.")
)

type planilhaErro string

func (e planilhaErro) Error() string { return string(e) }

// celulaData lê o valor BRUTO da célula (serial do Excel) quando é data de verdade, com
// fallback pra texto BR/ISO — a planilha de RH mistura os dois formatos na mesma coluna
// porque parte foi digitada à mão (mesmo motivo do comentário em paraData de funcionarios.ts).
func celulaData(f *excelize.File, sheet string, col, linha int) *time.Time {
	ref, _ := excelize.CoordinatesToCellName(col, linha)
	bruto, err := f.GetCellValue(sheet, ref, excelize.Options{RawCellValue: true})
	if err == nil && bruto != "" {
		if serial, err := strconv.ParseFloat(bruto, 64); err == nil {
			if t, err := excelize.ExcelDateToTime(serial, false); err == nil {
				t = time.Date(t.Year(), t.Month(), t.Day(), 0, 0, 0, 0, time.UTC)
				return &t
			}
		}
	}

	formatado, err := f.GetCellValue(sheet, ref)
	if err != nil {
		return nil
	}
	return parseDataTexto(formatado)
}

func parseDataTexto(texto string) *time.Time {
	texto = strings.TrimSpace(texto)
	if texto == "" {
		return nil
	}

	if t := parseDataBR(texto); t != nil {
		return t
	}
	if len(texto) >= 10 {
		if t, err := time.Parse("2006-01-02", texto[:10]); err == nil {
			return &t
		}
	}
	return nil
}

func parseDataBR(texto string) *time.Time {
	separador := ""
	for _, s := range []string{"/", "-", "."} {
		if strings.Contains(texto, s) {
			separador = s
			break
		}
	}
	if separador == "" {
		return nil
	}
	partes := strings.Split(texto, separador)
	if len(partes) != 3 {
		return nil
	}
	dia, err1 := strconv.Atoi(partes[0])
	mes, err2 := strconv.Atoi(partes[1])
	ano, err3 := strconv.Atoi(partes[2])
	if err1 != nil || err2 != nil || err3 != nil {
		return nil
	}
	if ano < 100 {
		ano += 2000
	}
	if mes < 1 || mes > 12 || dia < 1 || dia > 31 {
		return nil
	}
	t := time.Date(ano, time.Month(mes), dia, 0, 0, 0, 0, time.UTC)
	return &t
}

func celulaNumero(texto string) *float64 {
	texto = strings.TrimSpace(texto)
	if texto == "" {
		return nil
	}
	limpo := strings.NewReplacer("R$", "", " ", "", ".", "", ",", ".").Replace(texto)
	n, err := strconv.ParseFloat(limpo, 64)
	if err != nil {
		return nil
	}
	return &n
}
